crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -0,0 +1,379 @@
+    /// Any existing edge from A in the provided direction will be overwritten.
+    ///
+    /// The reverse edge will not be added, so navigation will only be possible in one direction.
+    /// If you want to add a symmetrical edge, use [`add_symmetrical_edge`](Self::add_symmetrical_edge) instead.
+    pub fn add_edge(&mut self, a: Entity, b: Entity, direction: CompassOctant) {
+        self.neighbors
+            .entry(a)
+            .or_insert(NavNeighbors::EMPTY)
+            .set(direction, b);
+    }
+
+    /// Adds a symmetrical edge between two entities in the navigation map.
+    /// The A -> B path will use the provided direction, while B -> A will use the [`CompassOctant::opposite`] variant.
+    ///
+    /// Any existing connections between the two entities will be overwritten.
+    pub fn add_symmetrical_edge(&mut self, a: Entity, b: Entity, direction: CompassOctant) {
+        self.add_edge(a, b, direction);
+        self.add_edge(b, a, direction.opposite());
+    }
+
+    /// Add symmetrical edges between all entities in the provided slice, looping back to the first entity at the end.
+    ///
+    /// This is useful for creating a circular navigation path between a set of entities, such as a menu.
+    pub fn add_looping_edges(&mut self, entities: &[Entity], direction: CompassOctant) {
+        for i in 0..entities.len() {
+            let a = entities[i];
+            let b = entities[(i + 1) % entities.len()];
+            self.add_symmetrical_edge(a, b, direction);
+        }
+    }
+
+    /// Gets the entity in a given direction from the current focus, if any.
+    pub fn get_neighbor(&self, focus: Entity, octant: CompassOctant) -> Option<Entity> {
+        self.neighbors
+            .get(&focus)
+            .and_then(|neighbors| neighbors.get(octant))
+    }
+
+    /// Looks up the neighbors of a given entity.
+    ///
+    /// If the entity is not in the map, [`None`] will be returned.
+    /// Note that the set of neighbors is not guaranteed to be non-empty though!
+    pub fn get_neighbors(&self, entity: Entity) -> Option<&NavNeighbors> {
+        self.neighbors.get(&entity)
+    }
+}
+
+/// A system parameter for navigating between focusable entities in a directional way.
+#[derive(SystemParam, Debug)]
+pub struct DirectionalNavigation<'w> {
+    /// The currently focused entity.
+    pub focus: ResMut<'w, InputFocus>,
+    /// The navigation map containing the connections between entities.
+    pub map: Res<'w, DirectionalNavigationMap>,
+}
+
+impl DirectionalNavigation<'_> {
+    /// Navigates to the neighbor in a given direction from the current focus, if any.
+    ///
+    /// Returns the new focus if successful.
+    /// Returns an error if there is no focus set or if there is no neighbor in the requested direction.
+    ///
+    /// If the result was `Ok`, the [`InputFocus`] resource is updated to the new focus as part of this method call.
+    pub fn navigate(
```

Comment:
This is for another PR, but I wonder if we should also consider "adjacent octants". It may be confusing to users if they accidentally pressed up-and-left on a UI with just a left and their action had no effect. I recognize this is ambiguous (what if you have both a left and an up but no top-left), but I think picking any direction is better than doing nothing.